### PR TITLE
core: Add support for N3DS memory mappings

### PR DESCRIPTION
**Pull request by FearlessTobi**
This adds support for the N3DS memory mappings.
It is enough for quite a few N3DS exclusives to go ingame now, including Xenoblade Chronicles 3D.

This code is a few months already and was written with the help of @wwylele. 
It is hence labeled WIP, and cause it still needs a lot of testing either way.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5103)

### CODE REVIEW

File: src/core/hle/kernel/memory.cpp [hunk 1]
```diff
@@ -55,14 +75,12 @@ void KernelSystem::MemoryInit(u32 mem_type) {
     }
 
     // We must've allocated the entire FCRAM by the end
-    ASSERT(base == Memory::FCRAM_SIZE);
+    ASSERT(base == is_new_3ds ? Memory::FCRAM_N3DS_SIZE : Memory::FCRAM_SIZE);
```
Comment (wwylele) [inline src/core/hle/kernel/memory.cpp]: The operation order is likely wrong here. It should be
```suggestion
    ASSERT(base == (is_new_3ds ? Memory::FCRAM_N3DS_SIZE : Memory::FCRAM_SIZE));
```
ASSERT(base == is_new_3ds ? Memory::FCRAM_N3DS_SIZE : Memory::FCRAM_SIZE);

**Comment by wwylele** [comment]
Also you need to update the unit test call to the kernel constructor. Other than these LGTM

File: src/core/hle/kernel/memory.cpp [hunk 2]
```diff
@@ -40,11 +41,30 @@ static const u32 memory_region_sizes[8][3] = {
...
 };
 
-void KernelSystem::MemoryInit(u32 mem_type) {
-    // TODO(yuriks): On the n3DS, all o3DS configurations (<=5) are forced to 6 instead.
-    ASSERT_MSG(mem_type <= 5, "New 3DS memory configuration aren't supported yet!");
+enum N3DSMode : u8 {
+    Mode6 = 1,
+    Mode7 = 2,
+    Mode6_2 = 3,
+};
```
Comment (zhaowenlan1779) [inline src/core/hle/kernel/memory.cpp]: I would suggest putting this in a namespace so the enum values have a scope.

File: src/core/loader/loader.h [hunk 3]
```diff
@@ -105,13 +105,22 @@ class AppLoader : NonCopyable {
...
         return std::make_pair(2, ResultStatus::Success);
     }
 
+    /**
+     * Loads the N3ds mode that this application uses.
+     * It defaults to 1 (Mode6)  if it can't read the information.
+     * @returns A pair with the optional N3ds mode, and the status.
+     */
+    virtual std::pair<std::optional<u8>, ResultStatus> LoadKernelN3dsMode() {
+        return std::make_pair(1, ResultStatus::Success);
```
Comment (zhaowenlan1779) [inline src/core/loader/loader.h]: Should probably be 0?

File: src/core/loader/loader.h [hunk 4]
```diff
@@ -105,13 +105,22 @@ class AppLoader : NonCopyable {
...
+     * @returns A pair with the optional system mode, and the status.
      */
     virtual std::pair<std::optional<u32>, ResultStatus> LoadKernelSystemMode() {
         // 96MB allocated to the application.
         return std::make_pair(2, ResultStatus::Success);
     }
 
+    /**
+     * Loads the N3ds mode that this application uses.
+     * It defaults to 1 (Mode6)  if it can't read the information.
```
Comment (zhaowenlan1779) [inline src/core/loader/loader.h]: This comment should be updated

**Comment by Timmy5942** [comment]
For some reason Fire Emblem Warriors still don't work, if its forced to memory 6 then it dose.

**Comment by B3n30** [comment]
It will be fixed as soon as we allow to change settings to n3ds mode.

**Comment by Timmy5942** [comment]
> It will be fixed as soon as we allow to change settings to n3ds mode.

Ok got ya, yes just as simple as the original idea made the game work
https://github.com/citra-emu/citra/compare/master...Timmy5942:N3DS_Memory_Boots_N3DS_Games?expand=1

Any good work , and hope you get it sorted then.

**Pull request merged**